src/main/scala/firrtl/transforms/DedupAnnotations.scala

```diff
@@ -0,0 +1,131 @@
+          // annotations will deduplicate into the same local annotation
+          case (left: MemoryRandomInitAnnotation, right: MemoryRandomInitAnnotation) =>
+            left.target.encapsulatingModule.equals(right.target.encapsulatingModule) &&
+              left.target.ref.equals(right.target.ref)
+          case (left: MemoryScalarInitAnnotation, right: MemoryScalarInitAnnotation) =>
+            val comparison = left.target.encapsulatingModule.equals(right.target.encapsulatingModule) &&
+              left.target.ref.equals(right.target.ref)
+
+            if (comparison && !left.value.equals(right.value))
+              // Error - deduplicated mem modules contain different scalar init values.
+              throw DifferingModuleAnnotationsException(left.target, right.target)
+
+            comparison
+          case (left: MemoryArrayInitAnnotation, right: MemoryArrayInitAnnotation) =>
+            val comparison = left.target.encapsulatingModule.equals(right.target.encapsulatingModule) &&
+              left.target.ref.equals(right.target.ref)
+
+            if (comparison && !left.values.equals(right.values))
+              // Error - deduplicated mem modules contain different array init values.
+              throw DifferingModuleAnnotationsException(left.target, right.target)
+
+            comparison
+          case (left: MemoryFileInlineAnnotation, right: MemoryFileInlineAnnotation) =>
+            val comparison = left.target.encapsulatingModule.equals(right.target.encapsulatingModule) &&
+              left.target.ref.equals(right.target.ref)
+
+            if (comparison && !left.hexOrBinary.equals(right.hexOrBinary))
+              // Error - deduplicated modules contain different binaries (should be the same).
+              throw DifferingModuleAnnotationsException(left.target, right.target)
+
+            comparison
+          case (_, _) => false
+        }
+      }
+    }
+  }
+
+  def dedupAnnotations(annotations: Seq[Annotation]): Seq[Annotation] = {
+    // Used to check all 'similar' annotations for deduplication. If two annotations
+    // would deduplicate to each other (have the same encapsulating module, and the same reference target),
+    // but otherwise have differing contents, then this would result in an error.
+    var annotationCheckGroups: Seq[Seq[Annotation]] = Seq(Seq())
+
+    // For every annotation, search for an appropriate group to append it to.
+    annotations.map(annotation => {
+      // If we previously added an annotation to the empty group, append a new empty group to the list
+      if (annotationCheckGroups.tail.nonEmpty) annotationCheckGroups = annotationCheckGroups :+ Seq()
+
+      // Append the annotation to the first matching group (this can be the empty group itself)
+      annotationCheckGroups = annotationCheckGroups.map(group =>
+        if (annotationBelongsInGroup(annotation, group)) group :+ annotation else group
+      )
+
```

Comment:
This logic is a little mindbending (and n^3 complexity where `n` is the number of annotations involved in this deduplication--to be fair, would be n^2 if it weren't for the list append on `annotationCheckGroups`). I'm also hoping that the code we end up here is easier to generalize.

I'd prefer an approach that is sort of in the spirit of [`groupBy`](https://www.scala-lang.org/api/2.12.13/scala/collection/Seq.html#groupBy[K](f:A=%3EK):scala.collection.immutable.Map[K,Repr]). I would first iterate on all annotations to see which ones "can dedup", then group them to see which "should dedup", then finally determine if they "do dedup". "do dedup" also needs an `InstanceGraph` to determine that an annotation applies to *all* instances of the module before actually doing the deduplication (otherwise it would be illegal), and we need a test for this, eg. 2 out of 3 instances have an annotation that dedups and the 3rd doesn't have an annotation at all, can't dedup in that case.

I'm thinking something like:

```scala
// Utility class for deduping annotations
// Deduped should == any other version of the same annotation 
// This type isn't quite right but it's something like this
private case class AnnoDedup[A, B <: Annotation](shouldDedupKey: A, deduped: B, orig: B)

// Does this utility already exist?
private def toLocal(rt: ReferenceTarget): ReferenceTarget =
  if (rt.isLocal) rt else rt.copy(module = rt.encapsulatingModule, rt.path = Nil)

// TODO: All of this can be folded into the annotation types
// May also apply to generic annotations: add *something* to the Annotation trait class
// Returns `Option` because it isn't implemented for all annotations
// Turns an Annotation into a Dedupable representation
def dedupable(annotation: Annotation): Option[AnnoDedup] = annotation match {
  // Example of what the implementation looks like for a particular annotation
  case a @ MemoryScalarInitAnnotation(target, value) =>
    val localTarget = toLocal(target)
    val shouldDedup = (localTarget, value)
    val deduped = a.copy(target = localTarget)
    AnnoDedup(shouldDedup, deduped, a)
  case _ => None
}

// ***** Step 1. Can Dedup *****
val resultAnnos = mutable.ArrayBuffer.empty[Annotation] // Annotations that aren't involved in the dedup
val canDedup = mutable.ArrayBuffer.empty[AnnoDedup]
annotations.foreach { anno =>
  dedupable(anno) match {
    case None => otherAnnos += anno
    case Some(d) => canDedup += d
  }
}

// ***** Step 2. Should Dedup *****
// This is a sketch, probably wrong
val groups = canDedup.groupBy(_.shouldDedupKey)
                     .mapValues(_.map(_._2)) // boilerplate due to lack of .groupMap in Scala 2.11 and 2.12

// ***** Step 3. Does Dedup *****
// This is a sketch, probably wrong, also need an InstanceGraph check
val groups.map { case (key, values) => 
  val deduped = values.map(_.deduped).distinct
  if (deduped.size == 1 && somehowCheckItsTheSameForAllInstancesOfTheSameModule) {
    resultAnnos += deduped.head
  } else {
    resultAnnos += values.map(_.orig)
  }
}
```